File: backend/src/main/java/com/kiwi/features/metrics/controllers/MetricsRepository.java
package com.kiwi.features.metrics.controllers;

import com.kiwi.features.metrics.data.MetricsPersistence;
import com.kiwi.features.users.data.UsersPersistence;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.time.LocalDate;
import java.util.Optional;

@Repository
public interface MetricsRepository extends JpaRepository<MetricsPersistence, Long> {
    Optional<MetricsPersistence> findByUserAndDate(UsersPersistence usersPersistence, LocalDate date);
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/metrics/MetricsRepository.kt
package com.bellako.kiwi.features.metrics

import com.bellako.kiwi.features.users.Email
import retrofit2.HttpException
import retrofit2.Response
import java.time.LocalDate

class MetricsRepository(private val api: IMetricsAPI) {
    suspend fun createMetrics(dto: MetricsDTO): Result<Unit> {
        return try {
            val response = api.createMetrics(dto)

            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                Result.failure(HttpException(response))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun updateMetrics(dto: MetricsDTO): Result<Unit> {
        return try {
            val response = api.updateMetrics(dto)

            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                Result.failure(HttpException(response))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getMetricsByDateAndUser(email: Email, date: LocalDate) : Result<MetricsDTO?> {
        return try {
            val response: Response<MetricsDTO> = api.getMetricsByDateAndUser(email.value, date.toString())
            if (response.isSuccessful) {
                Result.success(response.body())
            } else {
                Result.failure(HttpException(response))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}